File: python/tests/compatibility_wrapper.py
import sys
import json
import argparse
import os

from encrypted_storage import EncryptedStorage, errors

def main():
    parser = argparse.ArgumentParser(description="Test-only compatibility wrapper")
    parser.add_argument("operation", choices=["write", "read", "update", "delete"])
    parser.add_argument("--db", required=True)
    parser.add_argument("--schema", required=True)
    parser.add_argument("--content-type", dest="content_type", required=True)
    parser.add_argument("--object-uuid", dest="object_uuid", required=False) # Only needed for read but passed from write

    args = parser.parse_args()

    passphrase = os.environ.get("VAULT_PASSPHRASE")
    if not passphrase:
        print(json.dumps({"ok": False, "operation": args.operation, "error": "VAULT_PASSPHRASE environment variable is required"}))
        sys.exit(1)

    try:
        storage = EncryptedStorage(args.db)

        if args.operation == "write":
            if storage.get_status() == "uninitialized":
                storage.initialize_database(passphrase, "linux") # Using "linux" as dummy platform for tests
            else:
                storage.unlock_database(passphrase)

            payload_str = sys.stdin.read()
            payload = json.loads(payload_str)

            object_uuid = storage.store_payload(args.schema, args.content_type, payload)

            storage.close()

            print(json.dumps({
                "ok": True,
                "operation": "write",
                "object_uuid": object_uuid
            }))

        elif args.operation == "read":
            if not args.object_uuid:
                print(json.dumps({"ok": False, "operation": "read", "error": "--object-uuid is required for read operation"}))
                sys.exit(1)

            storage.unlock_database(passphrase)
            payload = storage.retrieve_payload(args.object_uuid)
            storage.close()

            print(json.dumps({
                "ok": True,
                "operation": "read",
                "payload": payload
            }))


        elif args.operation == "update":
            if not args.object_uuid:
                print(json.dumps({"ok": False, "operation": "update", "error": "--object-uuid is required for update operation"}))
                sys.exit(1)

            storage.unlock_database(passphrase)
            payload_str = sys.stdin.read()
            payload = json.loads(payload_str)
            storage.update_payload(args.object_uuid, args.schema, args.content_type, payload)
            storage.close()

            print(json.dumps({
                "ok": True,
                "operation": "update"
            }))

        elif args.operation == "delete":
            if not args.object_uuid:
                print(json.dumps({"ok": False, "operation": "delete", "error": "--object-uuid is required for delete operation"}))
                sys.exit(1)

            storage.unlock_database(passphrase)
            storage.delete_payload(args.object_uuid)
            storage.close()

            print(json.dumps({
                "ok": True,
                "operation": "delete"
            }))

    except errors.StorageError as e:
        print(json.dumps({
            "ok": False,
            "operation": args.operation,
            "error": str(e),
            "error_class": type(e).__name__
        }))
        sys.exit(1)
    except Exception as e:
        print(json.dumps({
            "ok": False,
            "operation": args.operation,
            "error": str(e)
        }))
        sys.exit(1)

if __name__ == "__main__":
    main()

File: nodejs/test/compatibility_wrapper.js
const fs = require('fs');
const EncryptedStorage = require('../src/storage');
const errors = require('../src/errors');

async function main() {
    const args = process.argv.slice(2);
    let operation = null;
    let dbPath = null;
    let schemaUuid = null;
    let contentType = null;
    let objectUuid = null;

    if (args.length > 0 && !args[0].startsWith('--')) {
        operation = args[0];
    }

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--db' && i + 1 < args.length) dbPath = args[++i];
        if (args[i] === '--schema' && i + 1 < args.length) schemaUuid = args[++i];
        if (args[i] === '--content-type' && i + 1 < args.length) contentType = args[++i];
        if (args[i] === '--object-uuid' && i + 1 < args.length) objectUuid = args[++i];
    }

    if (!operation || (operation !== 'write' && operation !== 'read') || !dbPath || !schemaUuid || !contentType) {
        console.log(JSON.stringify({ok: false, operation: operation || "unknown", error: "Missing required arguments"}));
        process.exit(1);
    }

    const passphrase = process.env.VAULT_PASSPHRASE;
    if (!passphrase) {
        console.log(JSON.stringify({ok: false, operation, error: "VAULT_PASSPHRASE environment variable is required"}));
        process.exit(1);
    }

    try {
        const storage = new EncryptedStorage(dbPath);

        if (operation === 'write') {
            const status = storage.getStatus();
            if (status === 'uninitialized') {
                await storage.initializeDatabase(passphrase, "linux"); // Using dummy platform
            } else {
                await storage.unlockDatabase(passphrase);
            }

            const payloadStr = fs.readFileSync(0, 'utf-8');
            let payload;
            try {
                payload = JSON.parse(payloadStr);
            } catch(e) {
                console.log(JSON.stringify({ok: false, operation, error: "Invalid JSON payload"}));
                process.exit(1);
            }

            const storedObjectUuid = storage.storePayload(schemaUuid, contentType, payload);
            storage.close();

            console.log(JSON.stringify({
                ok: true,
                operation: "write",
                object_uuid: storedObjectUuid
            }));

        } else if (operation === 'read') {
            if (!objectUuid) {
                console.log(JSON.stringify({ok: false, operation, error: "--object-uuid is required for read operation"}));
                process.exit(1);
            }

            await storage.unlockDatabase(passphrase);
            const payload = storage.retrievePayload(objectUuid);
            storage.close();

            console.log(JSON.stringify({
                ok: true,
                operation: "read",
                payload: payload
            }));
        }

    } catch (e) {
        console.log(JSON.stringify({
            ok: false,
            operation,
            error: e.message || String(e)
        }));
        process.exit(1);
    }
}

main().catch(err => {
    console.log(JSON.stringify({
        ok: false,
        operation: "unknown",
        error: String(err)
    }));
    process.exit(1);
});
